server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalance.java

```diff
@@ -19,8 +20,18 @@
  * The desired balance of the cluster, indicating which nodes should hold a copy of each shard.
  *
  * @param assignments a set of the (persistent) node IDs to which each {@link ShardId} should be allocated
+ * @param weightsPerNode The node weights calculated based on
+ * {@link org.elasticsearch.cluster.routing.allocation.allocator.BalancedShardsAllocator.WeightFunction#nodeWeight}
  */
-public record DesiredBalance(long lastConvergedIndex, Map<ShardId, ShardAssignment> assignments) {
+public record DesiredBalance(
+    long lastConvergedIndex,
+    Map<ShardId, ShardAssignment> assignments,
+    Map<DiscoveryNode, DesiredBalanceMetrics.NodeWeightStats> weightsPerNode
```

Comment:
It looks like this field is added to pass `weightsPerNode` from computer to reconciler to record stats there.

Why not record them to metrics in computer instead?
This way we do not need to pass this information nor keep it in memory until next allocation cycle.

Reply:
I think the difference is that in computer we might publish stats that we never reconciled towards because it might be superseded by a new computation. I thought this makes sure we publish stats for a desired balance we have submitted a reconcile task for.

Reply:
>I thought this makes sure we publish stats for a desired balance we have submitted a reconcile task for.

This is true, however this does not guarantee that completely reconciled those.
It is very possible that we saw them once, reconciled 1 or 2 shards out of the entire diff before receiving a new balance.

Reply:
yes, but to me it seems at least those are `DesiredBalance` instances that we have actually acted on. Publishing from computer might publish some weights that we have not actually acted on. I think also conceptually the weight stats are not that irrelevant to the `DesiredBalance`. Doing all of this in Reconciler also makes reusing `DesiredBalanceMetrics` simpler.

Reply:
It is not entirely clear what this metric represents. It is a quality of computed balance, but only if it was picked up by reconciled, but not necessarily fully reconciled. It looks more driven by the specifics of implementation rather than by clarity of what it represents.

I believe we should instead compute those metrics on top of RoutingNodes.
We can do it in the end of
(1) Reconciliation. This would represent the quality of allocation that is currently achieved in the cluster. I believe this would correctly highlight the clusters that might have computed good balance but are failing to reconcile it as it requires to many shard movements.
(2) (optionally) Computation. This would allow us to see the quality (if we are interested) of the allocation as it is computed according to the latest cluster state.

Reply:
ok we discussed elsewhere with Ievgen. We clarified that the metrics exposed here are just the weights and co that have resulted in the computed desired balance (that we start to reconcile towards), i.e., the details of the final state that the balancer considers to be desired. 

What Ievgen was asking for (based on `RoutingNodes`) is that once we know the goal that we are reconciling towards, we should also publish metrics for each middle state since it might take multiple reconciliations to arrive at the desired balance. That is something I will do in ES-9873. And the combination of these two could show what the goal is and at each specific time and how far away we are from that goal (based on the three weight components exposed here vs. their counterpart metrics that will be exposed after each reconciliation based on `RoutingNodes`, `ClusterState`, etc. similar to what is done in `AllocationStatsService`).